panic in TestRollingPolicy_Add

## Flaky Test

### Which jobs are failing
```
2023-11-15T17:04:55.9143408Z --- FAIL: TestRollingPolicy_Add (0.30s)
2023-11-15T17:04:55.9145312Z     --- FAIL: TestRollingPolicy_Add/test_policy_add#01 (0.10s)
2023-11-15T17:04:55.9147043Z panic: runtime error: index out of range [0] with length 0 [recovered]
2023-11-15T17:04:55.9148402Z 	panic: runtime error: index out of range [0] with length 0
2023-11-15T17:04:55.9149534Z 
2023-11-15T17:04:55.9150160Z goroutine 34 [running]:
2023-11-15T17:04:55.9151067Z testing.tRunner.func1.2({0x720020, 0xc00011c000})
2023-11-15T17:04:55.9152656Z 	/opt/hostedtoolcache/go/1.21.3/x64/src/testing/testing.go:1545 +0x366
2023-11-15T17:04:55.9154041Z testing.tRunner.func1()
2023-11-15T17:04:55.9155164Z 	/opt/hostedtoolcache/go/1.21.3/x64/src/testing/testing.go:1548 +0x630
2023-11-15T17:04:55.9156303Z panic({0x720020?, 0xc00011c000?})
2023-11-15T17:04:55.9157696Z 	/opt/hostedtoolcache/go/1.21.3/x64/src/runtime/panic.go:920 +0x270
2023-11-15T17:04:55.9159081Z github.com/tikv/pd/pkg/window.TestRollingPolicy_Add.func1(0x0?)
2023-11-15T17:04:55.9160653Z 	/home/runner/work/pd/pd/pkg/window/policy_test.go:70 +0x4af
2023-11-15T17:04:55.9161925Z testing.tRunner(0xc000222000, 0xc000216000)
2023-11-15T17:04:55.9163123Z 	/opt/hostedtoolcache/go/1.21.3/x64/src/testing/testing.go:1595 +0x239
2023-11-15T17:04:55.9164267Z created by testing.(*T).Run in goroutine 3
2023-11-15T17:04:55.9165589Z 	/opt/hostedtoolcache/go/1.21.3/x64/src/testing/testing.go:1648 +0x82b
2023-11-15T17:04:55.9166765Z FAIL	github.com/tikv/pd/pkg/window	8.319s
```

### CI link

https://github.com/lhy1024/pd/actions/runs/6880527737/job/18714932638
### Reason for failure (if possible)


### Anything else

--- a/pkg/window/policy_test.go
+++ b/pkg/window/policy_test.go
@@ -81,10 +81,10 @@ func TestRollingPolicy_Add(t *testing.T) {
 				if asExpected {
 					re.Less(math.Abs(point-policy.window.buckets[offset].Points[0]), 1e-6,
 						fmt.Sprintf("error, time since last append: %vms, last offset: %v", totalTS, lastOffset))
-					lastOffset = offset
 				}
 				re.Less(math.Abs(points[i]-policy.window.buckets[offset].Points[0]), 1e-6,
 					fmt.Sprintf("error, time since last append: %vms, last offset: %v", totalTS, lastOffset))
+				lastOffset = offset
 			}
 		})
 	}

--- a/pkg/window/policy_test.go
+++ b/pkg/window/policy_test.go
@@ -26,9 +26,12 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func GetRollingPolicy() *RollingPolicy {
-	w := NewWindow(Options{Size: 3})
-	return NewRollingPolicy(w, RollingPolicyOpts{BucketDuration: 100 * time.Millisecond})
+const defaultBucketDuration = 100 * time.Millisecond
+const defaultSize = 3
+
+func getRollingPolicy() *RollingPolicy {
+	w := NewWindow(Options{Size: defaultSize})
+	return NewRollingPolicy(w, RollingPolicyOpts{BucketDuration: defaultBucketDuration})
 }
 
 func TestRollingPolicy_Add(t *testing.T) {
@@ -45,6 +48,7 @@ func TestRollingPolicy_Add(t *testing.T) {
 			points:    []float64{1, 1},
 		},
 		{
+			// In CI, the actual sleep time may be more than 100 (timeSleep = 94).
 			timeSleep: []int{94, 250},
 			offset:    []int{0, 0},
 			points:    []float64{1, 1},
@@ -60,16 +64,23 @@ func TestRollingPolicy_Add(t *testing.T) {
 		t.Run("test policy add", func(t *testing.T) {
 			var totalTS, lastOffset int
 			timeSleep := test.timeSleep
-			policy := GetRollingPolicy()
+			beginTime := time.Now()
+			policy := getRollingPolicy()
+			points := make([]float64, defaultSize)
 			for i, n := range timeSleep {
 				totalTS += n
 				time.Sleep(time.Duration(n) * time.Millisecond)
-				offset, point := test.offset[i], test.points[i]
+				point := test.points[i]
+				offset := int(time.Since(beginTime)/defaultBucketDuration) % defaultSize
+				points[i] += point
 				policy.Add(point)
-
-				re.Less(math.Abs(point-policy.window.buckets[offset].Points[0]), 1e-6,
+				if offset == test.offset[i] {
+					re.Less(math.Abs(point-policy.window.buckets[offset].Points[0]), 1e-6,
+						fmt.Sprintf("error, time since last append: %vms, last offset: %v", totalTS, lastOffset))
+					lastOffset = offset
+				}
+				re.Less(math.Abs(points[i]-policy.window.buckets[offset].Points[0]), 1e-6,
 					fmt.Sprintf("error, time since last append: %vms, last offset: %v", totalTS, lastOffset))
-				lastOffset = offset
 			}
 		})
 	}
@@ -78,7 +89,7 @@ func TestRollingPolicy_Add(t *testing.T) {
 func TestRollingPolicy_AddWithTimespan(t *testing.T) {
 	re := require.New(t)
 	t.Run("timespan < bucket number", func(t *testing.T) {
-		policy := GetRollingPolicy()
+		policy := getRollingPolicy()
 		// bucket 0
 		policy.Add(0)
 		// bucket 1
@@ -102,7 +113,7 @@ func TestRollingPolicy_AddWithTimespan(t *testing.T) {
 	})
 
 	t.Run("timespan > bucket number", func(t *testing.T) {
-		policy := GetRollingPolicy()
+		policy := getRollingPolicy()
 
 		// bucket 0
 		policy.Add(0)

--- a/pkg/window/policy_test.go
+++ b/pkg/window/policy_test.go
@@ -67,6 +67,7 @@ func TestRollingPolicy_Add(t *testing.T) {
 			beginTime := time.Now()
 			policy := getRollingPolicy()
 			points := make([]float64, defaultSize)
+			asExpected := true
 			for i, n := range timeSleep {
 				totalTS += n
 				time.Sleep(time.Duration(n) * time.Millisecond)
@@ -74,7 +75,10 @@ func TestRollingPolicy_Add(t *testing.T) {
 				offset := int(time.Since(beginTime)/defaultBucketDuration) % defaultSize
 				points[i] += point
 				policy.Add(point)
-				if offset == test.offset[i] {
+				if offset != test.offset[i] {
+					asExpected = false
+				}
+				if asExpected {
 					re.Less(math.Abs(point-policy.window.buckets[offset].Points[0]), 1e-6,
 						fmt.Sprintf("error, time since last append: %vms, last offset: %v", totalTS, lastOffset))
 					lastOffset = offset

